### PR TITLE
wip: new post about bucklescript ffi

### PR DESCRIPTION
**Pull request by bobzhang**

### CODE REVIEW

File: website/blog/2019-05-21-ffi-overview.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
+---
+title: A high level overview of BuckleScript interop with Javascript
+---
+
+When user starts to use BuckleScript to develop applications on JS platform, they have to interop with various APIs provided by JS platform. 
+
+In theory, like [Elm](https://elm-lang.org/), bucklescript could ship a comperhensive library which contains what most people would like to use daily, but this is in particular 
+challenging given JS is running on so many platforms for example, [Electron](https://electronjs.org/), [Node](https://nodejs.org/) and Browser,  yet eatch platform is still evolving quickly. So we have to provide a mechanism to allow users to binding to native JS API quickly in the userland.
```
Comment (ulrikstrid) [inline website/blog/2019-05-21-ffi-overview.md]: `,  yet eatch` -> `, yet each`
`users to binding to native` -> `users to bind to native`

File: website/blog/2019-05-21-ffi-overview.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+There are lots of trade-off when design such a FFI bridge between OCaml and JavaScript API, below we list a few key items which we think has an important impact on our design.
```
Comment (ulrikstrid) [inline website/blog/2019-05-21-ffi-overview.md:10]: `design` -> `designing`

Comment (yawaramin) [reply website/blog/2019-05-21-ffi-overview.md:10]: Design seems fine here.

File: website/blog/2019-05-21-ffi-overview.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+Note that a nice thing about such mechanism is that no separate JS file is needed so no change to the build system is need in most cases, using such mechanism, BuckleScript users can already deal with most bindings.
+
+## Interop via attributes
+
+If you are a developer busy shipping, the mechanism above would cover almost all you need. A minor disadvange of such mechanism is that it comes with a cost, such raw function can not be inlined since it is JavaScript that BuckleScript compiler does not have a deep knowledge.
+
+We are going through such mechanism via a small example of binding to JS date, there are lots of advanced topics in the [documentation](https://bucklescript.github.io/docs/en/interop-overview), we are only talking about one of the mostly used method here.
+
+The key idea is to bind your JS object as [an abstract data type](https://en.wikipedia.org/wiki/Abstract_data_type) where a data type is defined by its behavior from the point of view of a user  of the data instead of concrete representations.
```
Comment (yawaramin) [inline website/blog/2019-05-21-ffi-overview.md:76]: Great explanation!

File: website/blog/2019-05-21-ffi-overview.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+## Interop design constraints
+
+- BuckleScript is still OCaml
+
+    We are not inventing a new language, in particular, we can not change the concrete syntax of OCaml. Luckily, OCaml introduced [attributes](https://caml.inria.fr/pub/docs/manual-ocaml/extn.html#sec260) and [extension nodes](https://caml.inria.fr/pub/docs/manual-ocaml/extn.html#sec262) since 4.02, which allows us to customize the language in a minor style. To be a good citizen in the OCaml community, all attributes introduced by BuckleScript are prefixed with `bs`.
+
+- Bare metal efficiency should always be possible for experts in pure OCaml
+
+    Efficency is at the heart of BuckleScript's design philosophy, in terms of both compilation speed and runtime performance. There are also other strongly typed functional languages running on JS platform before we made BuckleScript, one thing in particular confused me is that in those languages, people have to write `native JS` to gain performance. Our goal is that when performance really matters, it is still possible for experts to write pure OCaml without digging into `native JS`, users don't have to make a choice between performance and type safety.
```
Comment (yawaramin) [inline website/blog/2019-05-21-ffi-overview.md]: Efficency -> Efficiency

File: website/blog/2019-05-21-ffi-overview.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
+---
+title: A high level overview of BuckleScript interop with Javascript
+---
+
+When user starts to use BuckleScript to develop applications on JS platform, they have to interop with various APIs provided by JS platform. 
+
+In theory, like [Elm](https://elm-lang.org/), bucklescript could ship a comperhensive library which contains what most people would like to use daily, but this is in particular
```
Comment (yawaramin) [inline website/blog/2019-05-21-ffi-overview.md]: comperhensive -> comprehensive

**Review by yawaramin — COMMENTED** [review]
Great–I will heavily publicize it :-)

File: website/blog/2019-05-21-ffi-overview.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+If you are a developer busy shipping, the mechanism above would cover almost all you need. A minor disadvange of such mechanism is that it comes with a cost, such raw function can not be inlined since it is JavaScript that BuckleScript compiler does not have a deep knowledge.
+
+We are going through such mechanism via a small example of binding to JS date, there are lots of advanced topics in the [documentation](https://bucklescript.github.io/docs/en/interop-overview), we are only talking about one of the mostly used method here.
+
+The key idea is to bind your JS object as [an abstract data type](https://en.wikipedia.org/wiki/Abstract_data_type) where a data type is defined by its behavior from the point of view of a user  of the data instead of concrete representations.
+
+```ocaml
+type date
+external fromFloat : float -> date = "Date" [@@.new]
```
Comment (sync) [inline website/blog/2019-05-21-ffi-overview.md]: [@@bs.new]

Comment (sync) [inline website/blog/2019-05-21-ffi-overview.md]:
```suggestion
external fromFloat : float -> date = "Date" [@@bs.new]
```

**Review by sync — COMMENTED** [review]
Just a typo

**Comment by bobzhang** [comment]
Hi all, thanks, revised per suggestions

**Pull request merged**